src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/NuGetUIContext.cs

```diff
@@ -93,5 +115,80 @@ public async Task<IModalProgressDialogSession> StartModalProgressDialogAsync(str
+            ISettings settings,
+            IVsSolutionManager solutionManager,
+            IPackageRestoreManager packageRestoreManager,
+            IOptionsPageActivator optionsPageActivator,
+            IUserSettingsManager userSettingsManager,
+            IDeleteOnRestartManager deleteOnRestartManager,
+            IEnumerable<IVsPackageManagerProvider> packageManagerProviders,
+            INuGetLockService lockService,
+            CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(sourceRepositoryProvider);
+            Assumes.NotNull(settings);
+            Assumes.NotNull(solutionManager);
+            Assumes.NotNull(packageRestoreManager);
+            Assumes.NotNull(optionsPageActivator);
+            Assumes.NotNull(userSettingsManager);
+            Assumes.NotNull(deleteOnRestartManager);
+            Assumes.NotNull(packageManagerProviders);
+            Assumes.NotNull(lockService);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            IServiceBroker serviceBroker = await BrokeredServicesUtilities.GetRemoteServiceBrokerAsync();
+
+            var solutionManagerService = new NuGetSolutionManagerServiceWrapper()
+            {
+                Service = await GetSolutionManagerServiceAsync(cancellationToken)
+            };
+
+            var packageManager = new NuGetPackageManager(
+                sourceRepositoryProvider,
+                settings,
+                solutionManager,
+                deleteOnRestartManager);
+
+            var actionEngine = new UIActionEngine(
+                sourceRepositoryProvider,
+                packageManager,
+                lockService);
+
+            return new NuGetUIContext(
+                sourceRepositoryProvider,
+                serviceBroker,
+                solutionManager,
+                solutionManagerService,
+                packageManager,
+                actionEngine,
+                packageRestoreManager,
+                optionsPageActivator,
+                userSettingsManager,
+                packageManagerProviders);
+        }
+
+        private void OnAvailabilityChanged(object sender, BrokeredServicesChangedEventArgs e)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    _solutionManagerService.Service = await GetSolutionManagerServiceAsync(CancellationToken.None);
+                })
+                .PostOnFailure(nameof(NuGetUIContext), nameof(OnAvailabilityChanged));
+        }
+
+        private static async ValueTask<INuGetSolutionManagerService> GetSolutionManagerServiceAsync(CancellationToken cancellationToken)
+        {
+            IServiceBroker serviceBroker = await BrokeredServicesUtilities.GetRemoteServiceBrokerAsync();
+
+#pragma warning disable ISB001 // Dispose of proxies
+            return await serviceBroker.GetProxyAsync<INuGetSolutionManagerService>(
```

Comment:
>GetProxyAsync [](start = 39, length = 13)

How will we handle null here? I know you check for null in the set but do we want to throw instead of using the NullSolutionManagerService?

Reply:
I think we should throw.